src/core/evaluator.js

```diff
@@ -2202,91 +2256,257 @@ class PartialEvaluator {
+    function compareWithLastPosition(fontSize) {
+      if (
+        !combineTextItems ||
+        !textState.font ||
+        !textContentItem.prevTransform
+      ) {
+        return;
+      }
+
+      const currentTransform = getCurrentTextTransform();
+      const posX = currentTransform[4];
+      const posY = currentTransform[5];
+      const lastPosX = textContentItem.prevTransform[4];
+      const lastPosY = textContentItem.prevTransform[5];
+
+      if (lastPosX === posX && lastPosY === posY) {
+        return;
+      }
+
+      const advanceX = (posX - lastPosX) / textContentItem.textAdvanceScale;
+      const advanceY = (posY - lastPosY) / textContentItem.textAdvanceScale;
+      const HALF_LAST_CHAR = -0.5 * textContentItem.lastCharSize;
+
+      if (textState.font.vertical) {
+        if (
+          Math.abs(advanceX) <
+            textContentItem.prevFontSize * 1000 /* same column */ &&
+          HALF_LAST_CHAR <= advanceY
+        ) {
+          if (advanceY > textContentItem.trackingSpaceMin) {
+            textContentItem.height += advanceY;
+          } else if (
+            !addFakeSpaces(advanceY, 0, textContentItem.prevTransform)
+          ) {
+            if (textContentItem.str.length === 0) {
+              textContent.items.push({
+                str: " ",
+                dir: "ltr",
+                width: 0,
+                height: advanceY,
+                transform: textContentItem.prevTransform,
+                fontName: textContentItem.fontName,
+                isSpace: true,
+                hasEOL: false,
+              });
+            } else {
+              textContentItem.height += advanceY;
+            }
+          }
         } else {
-          glyphWidth = glyph.width;
+          appendEOL();
         }
+      } else if (
+        Math.abs(advanceY) <
+          textContentItem.prevFontSize * 1000 /* same line */ &&
+        HALF_LAST_CHAR <= advanceX
+      ) {
+        if (advanceX <= textContentItem.trackingSpaceMin) {
+          textContentItem.width += advanceX;
+        } else if (!addFakeSpaces(advanceX, 0, textContentItem.prevTransform)) {
+          if (textContentItem.str.length === 0) {
+            textContent.items.push({
+              str: " ",
+              dir: "ltr",
+              width: advanceX,
+              height: 0,
+              transform: textContentItem.prevTransform,
+              fontName: textContentItem.fontName,
+              isSpace: true,
+              hasEOL: false,
+            });
+          } else {
+            textContentItem.width += advanceX;
+          }
+        }
+      } else {
+        appendEOL();
+      }
+    }
 
-        var glyphUnicode = glyph.unicode;
-        var NormalizedUnicodes = getNormalizedUnicodes();
-        if (NormalizedUnicodes[glyphUnicode] !== undefined) {
-          glyphUnicode = NormalizedUnicodes[glyphUnicode];
+    function buildTextContentItem(chars, extraSpacing, isFirstChunk) {
```

Comment:
Nit: You may want to consider changing this function to take a parameter object instead, which improves the self-documention at the call-sites, since some of the calls below are a little hard read as-is.
(That way you could also make use of default values, removing the need to pass all parameters everywhere.)

One example is the  `buildTextContentItem(args[0], 0, true);` calls, where it's not really obvious what the arguments represent without going back to the function definition.